source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -218,12 +219,239 @@ export class ValidationResults {
     }
 }
 
+export abstract class PropertyDefinition {
+    abstract parse(value: any): any;
+    abstract toJSON(target: object, value: any);
+
+    constructor(
+        readonly targetVersion: Shared.TargetVersion,
+        readonly name: string,
+        readonly defaultValue?: any) { }
+}
+
+export abstract class TypedPropertyDefinition<T> extends PropertyDefinition {
+    constructor(
+        readonly targetVersion: Shared.TargetVersion,
+        readonly name: string,
+        readonly defaultValue?: T) {
+        super(targetVersion, name, defaultValue);
+    }
+}
+
+export class StringPropertyDefinition extends TypedPropertyDefinition<string> {
+    parse(value: any): any {
+        let parsedValue = Utils.getStringValue(value, this.defaultValue);
+        let isUndefined = parsedValue === undefined || (parsedValue === "" && this.treatEmptyAsUndefined);
+
+        if (!isUndefined && this.regEx !== undefined) {
+            let matches = this.regEx.exec(value);
+
+            if (!matches) {
+                // TODO: Log parse error
```

Comment:
>// TODO: Log parse error [](start = 16, length = 24)

We should use a tracking bug here (unless we're not actually going to log an error here, in which case the comment should be removed)

Reply:
Most (probably all) TODOs are addressed in later commits. These were placeholders.

Reply:
I see. Should I just review the end product rather than per-commit?

---
In reply to: [331170192](https://github.com/microsoft/AdaptiveCards/pull/3508#discussion_r331170192) [](ancestors = 331170192)

Reply:
No, you will still need to review every commit, sorry. But there will be multiple instances of outdated changes. GitHub shows an "outdated" tag if that helps.